hazelcast/src/main/java/com/hazelcast/util/HashUtil.java

```diff
@@ -111,28 +110,34 @@ public static int MurmurHash3_x86_32(byte[] data, int offset, int len, int seed)
         return h1;
     }
 
-    public static int MurmurHash3_x86_32_direct(long address, int offset, int len) {
-        return MurmurHash3_x86_32_direct(address, offset, len, DEFAULT_MURMUR_SEED);
+    public static int MurmurHash3_x86_32(byte[] data, int offset, int len) {
+        return MurmurHash3_x86_32(data, offset, len, DEFAULT_MURMUR_SEED);
     }
 
+
     /**
-     * Returns the MurmurHash3_x86_32 hash.
+     * Returns the MurmurHash3_x86_32 hash of a memory block accessed by the provided {@link MemoryAccessor}.
      */
-    public static int MurmurHash3_x86_32_direct(long address, int offset, int len, int seed) {
+    public static int MurmurHash3_x86_32_direct(MemoryAccessor mem, long base, int offset, int len, int seed) {
+
+        // (len & ~(MURMUR32_BLOCK_SIZE - 1)) is the length rounded down to the Murmur32 block size boundary
+        final long tailStart = base + offset + (len & ~(MURMUR32_BLOCK_SIZE - 1));
+
         int c1 = 0xcc9e2d51;
         int c2 = 0x1b873593;
 
         int h1 = seed;
-        // round down to 4 byte block
-        int roundedEnd = offset + (len & 0xfffffffc);
 
-        for (int i = offset; i < roundedEnd; i += 4) {
-            // little endian load order
-            int k1 = LITTLE_ENDIAN ? MEM.getInt(address + i)
-                    : (MEM.getByte(address + i) & 0xff)
-                    | ((MEM.getByte(address + i + 1) & 0xff) << 8)
-                    | ((MEM.getByte(address + i + 2) & 0xff) << 16)
-                    | (MEM.getByte(address + i + 3) << 24);
+        final boolean memIsLittleEndian = !mem.isBigEndian();
+
+        for (long blockAddr = base + offset; blockAddr < tailStart; blockAddr += MURMUR32_BLOCK_SIZE) {
+            // little-endian load order
+            int k1 = memIsLittleEndian
+                    ? mem.getInt(blockAddr)
```

Comment:
If the given `MemoryAccessor` is not alignment aware and if the `offset` (so `blockAddr`) is not 4 bytes aligned, this might cause `SIGBUS` on platforms which don't support unaligned memory access. If we will give this responsibility to the given `MemoryAccessor`, this should be stated in the javadoc.